core/services/gateway/handlers/functions/allowlist/orm.go

```diff
@@ -0,0 +1,123 @@
+package allowlist
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/pkg/errors"
+
+	"github.com/jmoiron/sqlx"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+)
+
+//go:generate mockery --quiet --name ORM --output ./mocks/ --case=underscore
+type ORM interface {
+	GetAllowedSenders(offset, limit uint, qopts ...pg.QOpt) ([]common.Address, error)
+	CreateAllowedSenders(allowedSenders []common.Address, qopts ...pg.QOpt) error
+	DeleteAllowedSenders(blockedSenders []common.Address, qopts ...pg.QOpt) error
+}
+
+type orm struct {
+	q                     pg.Q
+	lggr                  logger.Logger
+	routerContractAddress common.Address
+}
+
+var _ ORM = (*orm)(nil)
+var (
+	ErrInvalidParameters = errors.New("invalid parameters provided to create a functions contract cache ORM")
+)
+
+const (
+	tableName = "functions_allowlist"
+)
+
+func NewORM(db *sqlx.DB, lggr logger.Logger, cfg pg.QConfig, routerContractAddress common.Address) (ORM, error) {
+	if db == nil || cfg == nil || lggr == nil || routerContractAddress == (common.Address{}) {
+		return nil, ErrInvalidParameters
+	}
+
+	return &orm{
+		q:                     pg.NewQ(db, lggr, cfg),
+		lggr:                  lggr,
+		routerContractAddress: routerContractAddress,
+	}, nil
+}
+
+func (o *orm) GetAllowedSenders(offset, limit uint, qopts ...pg.QOpt) ([]common.Address, error) {
+	var addresses []common.Address
+	stmt := fmt.Sprintf(`
+		SELECT allowed_address
+		FROM %s
+		WHERE router_contract_address = $1
+		ORDER BY id ASC
+		OFFSET $2
+		LIMIT $3;
+	`, tableName)
+	err := o.q.WithOpts(qopts...).Select(&addresses, stmt, o.routerContractAddress, offset, limit)
+	if err != nil {
+		return addresses, err
+	}
+	o.lggr.Debugf("Successfully fetched allowed sender list from DB. offset: %d, limit: %d, length: %d", offset, limit, len(addresses))
+
+	return addresses, nil
+}
+
+func (o *orm) CreateAllowedSenders(allowedSender []common.Address, qopts ...pg.QOpt) error {
+	var valuesPlaceholder []string
+	for i := 1; i <= len(allowedSender)*2; i += 2 {
+		valuesPlaceholder = append(valuesPlaceholder, fmt.Sprintf("($%d, $%d)", i, i+1))
+	}
+
+	stmt := fmt.Sprintf(`
+		INSERT INTO %s (allowed_address, router_contract_address)
+		VALUES %s ON CONFLICT (allowed_address, router_contract_address) DO NOTHING;`, tableName, strings.Join(valuesPlaceholder, ", "))
+
+	var args []interface{}
+	for _, as := range allowedSender {
+		args = append(args, as, o.routerContractAddress)
+	}
+
+	_, err := o.q.WithOpts(qopts...).Exec(stmt, args...)
+	if err != nil {
+		return err
+	}
+
+	o.lggr.Debugf("Successfully stored allowed sender: %s for routerContractAddress: %s", allowedSender, o.routerContractAddress)
```

Comment:
How will "%s" work with an array?

Reply:
checking with the tests it parses it correctly 
`logger.go:130: 2024-01-24T10:52:24.474+0100	DEBUG	allowlist/orm.go:89	Successfully stored allowed senderssss: [0xa761778E3CA16558880B4bc16c2A2e652ed142F0 0xa761778E3CA16558880B4bc16c2A2e652ed142F0 0xa761778E3CA16558880B4bc16c2A2e652ed142F0 0xF1dBe1C701F0B33103C1ED0864F4C7a681294623] for routerContractAddress: 0x742B34d3a253878be65490f43dDeDD2BE40AD065`

but indeed `%v` would be more accurate, will change it 👍🏼